robot_skills/src/robot_skills/perception.py

```diff
@@ -184,11 +189,83 @@ def detect_faces(self, image=None, stamp=False):
         """
         if not image:
             image = self.get_image()
+        # ToDo: create a consistent API: either return a tuple, a message or a list
         if stamp:
             return self._get_faces(image).recognitions, image.header.stamp
         else:
             return self._get_faces(image).recognitions
 
+    def detect_operator_face(
+        self,
+        image: typing.Optional[Image] = None,
+        expected_operator_position: typing.Optional[kdl.Vector] = None,
+        operator_distance_threshold: typing.Optional[float] = 0.5,
+    ) -> Recognition:
+        """
+        Snap an image with the camera and return the detected face closest to the expected position.
+
+        N.B.: this seems more reliable than using the size of the ROI, which might be affected by, e.g., a big hairdo.
+
+        :param image: image to use for recognition
+        :param expected_operator_position: expected operator position w.r.t. robot base
+        :param operator_distance_threshold: people outside this radius from the expected position are discarded
+        :return: image_recognition_msgs/Recognition
+        :raises: RuntimeError
+        """
+        recognitions = self.detect_faces(image)
+        if not recognitions:
+            raise RuntimeError("No faces detected")
+        operator_recognition = self._filter_operator_recognition(
+            recognitions, expected_operator_position, operator_distance_threshold
+        )
+        return operator_recognition
+
+    def _filter_operator_recognition(
+        self,
+        recognitions: typing.List[Recognition],
+        expected_operator_pos: typing.Optional[kdl.Vector] = None,
+        threshold: float = 0.5,
+    ) -> Recognition:
+        """
+        For all provided recognitions, project the ROI and determine the one closest to the expected operator position,
+        computed on the floor plane.
+
+        :param recognitions: recognitions
+        :param expected_operator_pos: expected position of the operator w.r.t. the robot
+        :param threshold: if the distance between the recognition and the expected pos exceed the threshold, an error
+        is raised
+        :return: Recognition closed to the expected position
+        :raises: RuntimeError
+        """
+        frame_id = f"{self.robot_name}/base_link"
```

Comment:
The robot object can have different base link names. This is stored as robot.base_link_frame or something like that. Perhaps this should be added to the constructor of the perception part and stored as an attribute

Reply:
See 8904256a000980011173ba30f016b191dc1f8a79